commons/src/main/java/com/simpletools/calculator/commons/operations/OperationFactory.kt

```diff
@@ -2,20 +2,33 @@ package com.simpletools.calculator.commons.operations
 
 import com.simpletools.calculator.commons.operations.base.Operation
 import com.simpletools.calculator.commons.helpers.*
+import com.simpletools.calculator.commons.operations.base.BinaryOperation
+import com.simpletools.calculator.commons.operations.base.UnaryOperation
 
 object OperationFactory {
 
     fun forId(id: String, baseValue: Double, secondValue: Double): Operation? {
         when (id) {
-            PLUS -> return PlusOperation(baseValue, secondValue)
-            MINUS -> return MinusOperation(baseValue, secondValue)
-            DIVIDE -> return DivideOperation(baseValue, secondValue)
-            MULTIPLY -> return MultiplyOperation(baseValue, secondValue)
-            PERCENTAGE -> return PercentageOperation(baseValue)
-            POWER -> return PowerOperation(baseValue, secondValue)
-            ROOT -> return RootOperation(baseValue)
-            NEGATIVE -> return NegativeOperation(baseValue)
+            PLUS -> return binarySet(PlusOperation, baseValue, secondValue)
+            MINUS -> return binarySet(MinusOperation, baseValue, secondValue)
+            DIVIDE -> return binarySet(DivideOperation, baseValue, secondValue)
+            MULTIPLY -> return binarySet(MultiplyOperation, baseValue, secondValue)
+            PERCENTAGE -> return unarySet(PercentageOperation, baseValue)
+            POWER -> return binarySet(PowerOperation, baseValue, secondValue)
+            ROOT -> return unarySet(RootOperation, baseValue)
+            NEGATIVE -> return unarySet(NegativeOperation, baseValue)
             else -> return null
         }
     }
+
+    private fun binarySet(operation : BinaryOperation, baseValue : Double, secondValue : Double) : Operation? {
```

Comment:
Can you explain your reasoning behind passing the operation singletons as a parameter for these operations? Why this architectural change?

Reply:
These private functions are helpers that do the setting of `Operation` state variables for the `OperationFactory`. They prevent duplicate code inside `OperationFactory.forId()`. The reason the `Operation`s have their state set from the outside, instead of through the constructor as before, is because of a limitation preventing passing initialization arguments with the Kotlin `object` expression. I had found a workaround for passing initialization arguments to a Kotlin singleton [here](https://medium.com/@BladeCoder/kotlin-singletons-with-argument-194ef06edd9e) but it was more difficult to understand and implement.